pkg/auth/authenticator/token/cache/cache.go

```diff
@@ -0,0 +1,70 @@
+package cache
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/hashicorp/golang-lru"
+
+	kerrs "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/util"
+
+	"github.com/openshift/origin/pkg/auth/authenticator"
+)
+
+type CacheAuthenticator struct {
+	authenticator authenticator.Token
+
+	cache *lru.Cache
+
+	maxAge time.Duration
+	now    func() time.Time
+}
+
+type cacheRecord struct {
+	created time.Time
+	user    user.Info
+	ok      bool
+	err     error
+}
+
+// NewAuthenticator returns an authenticator that caches the results of the given authenticator
+func NewAuthenticator(a authenticator.Token, maxAge time.Duration, maxCount int) (authenticator.Token, error) {
+	cache, err := lru.New(maxCount)
+	if err != nil {
+		return nil, err
+	}
+	return &CacheAuthenticator{
+		authenticator: a,
+		cache:         cache,
+		maxAge:        maxAge,
+		now:           time.Now,
+	}, nil
+}
+
+func (c *CacheAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+	if value, hit := c.cache.Get(token); hit {
```

Comment:
This should be bumping the LRU count, but if its expired and you have a failure getting the authentication result, you don't remove the token.  That could cause other non-expired entries to lose their LRU priority.